File: DWD/ani_DWD_grid.m
%% Animation GRID
% animate gridded data and save as animated gif
clearvars
addpath data
% atmospheric data downloaded and processed with load_DWD_grid_data.m
load atmodata
% grid of geographical coordinates to be used instead of Gauß-Krueger
load coordlatlon
[m, n, ~]=size(adata);
lat=reshape(clatlon(:,1),m,n);
lon=reshape(clatlon(:,2),m,n);
%% Germany
% whole dataset
mtp=zeros(size(adata,1),size(adata,2));
adata2=zeros(size(adata));
% Mean 1961-1990
ind=find(zp>1960 & zp<1991);
for hi=1:size(adata,1)
    for h2=1:size(adata,2)
        if isnan(adata(hi,h2,1))
            mtp(hi,h2)=NaN;
            adata2(hi,h2,:)=NaN;
        else
            mtp(hi,h2)=mean(adata(hi,h2,ind));
            adata2(hi,h2,:)=adata(hi,h2,:)-mtp(hi,h2);
        end
    end
end
%% iterate over all years
% plot each year and save to single gif files
% figure is black with all grids and annotations removed
hh=figure('color',[0 0 0 ],'position',[100 100 600 1200]);
wm=worldmap([47 55.5],[5.8 15.5]);
setm(wm,'grid','off','frame','off');
set(findall(wm,'Tag','PLabel'),'visible','off')
set(findall(wm,'Tag','MLabel'),'visible','off')
% worldmap('germany')
load cmap3
filename='aniDWD.gif';
capturegif=true;
for hi=1:length(zp)
    hg=geoshow(lat,lon,adata2(:,:,hi),'displaytype','surface');
%     cb=colorbar;
%     set(get(cb,'Label'),'string','\delta Temperature / °C')
    colormap(cmap)
    caxis([-3 4.5])
    title(['\fontsize{20}{\color{white}' num2str(zp(hi,1)) '}']);
    drawnow
    if capturegif
        % Capture the plot as an image
        frame = getframe(hh);
        im = frame2im(frame);
        [imind,cm] = rgb2ind(im,256);
        % Write to the GIF File
        if hi == 1
            imwrite(imind,cm,filename,'gif', 'Loopcount',inf);
        else
            imwrite(imind,cm,filename,'gif','WriteMode','append');
        end
    end
    delete(hg)    
end

File: DWD/ani_DWD_grid.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
 Plot of annual mean temperature for Germany from DWD data and save as 
 individual plots and animated gif. 
"""
import scipy.io as sp
import numpy as np
import imageio
from matplotlib.colors import ListedColormap
import multiprocessing as mp
#import pyfunc.DWD_func as dwdf
#from pyfunc.DWD_func import plot_fun2

def myfun():
    from pyfunc.DWD_func import plot_fun2
    # atmospheric data downloaded and processed with load_DWD_grid_data.m
    h = np.load('atmodata.npz')
    #with np.load('atmodata.npz') as data:
    adata = h['a']
    zp = h['b']
    nds=len(zp)
    # dimension of grids from header of ascii file
    ncols = 654
    nrows = 866
    # grid of geographical coordinates to be used instead of Gauss-Krueger
    mat_contents = sp.loadmat('data/coordlatlon.mat')
    x = mat_contents.get("clatlon")
    lat=np.reshape(x[:,0],(ncols,nrows)).T
    lon=np.reshape(x[:,1],(ncols,nrows)).T
    # subtract mean temperature 1961-1990 for each data point
    ind = np.where((zp>1960.) & (zp<1991.))
    tepoch=adata[:,:,ind[0]]
    mtp = np.nanmean(tepoch,axis=2)
    # Colormap
    mat_contents = sp.loadmat('data/cmap3.mat')
    pcmap = mat_contents.get("cmap")
    mycmap = ListedColormap(pcmap)
    # generate plots with multiprocessing
    pnames=[]
    p = mp.Pool(mp.cpu_count())
    resi=[p.apply_async(plot_fun2, args = (lon, lat, adata[:,:,hi]-mtp,str(int(zp[hi])),mycmap)) 
               for hi in range(0,nds) ]
    p.close()
    p.join()
    for hi in range(0,nds):
        pname=resi[hi].get()
        pnames.append(pname)
    # save individual files as animated gif
    with imageio.get_writer('tempgif.gif', mode='I') as writer:
        for pname1 in pnames:
            image = imageio.imread(pname1)
            writer.append_data(image)

if __name__ == '__main__':
    myfun()

